### PR TITLE
Adapt `RemoteExternalities` and its related types to be used with generic hash parameters

### PR DESCRIPTION
**Pull request by lean-apple**
Closes  https://github.com/paritytech/polkadot-sdk/issues/3737

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5742279

**Comment by ggwpez** [comment]
bot fmt

File: substrate/utils/frame/remote-externalities/src/lib.rs [hunk 1]
```diff
@@ -63,21 +63,21 @@ const SNAPSHOT_VERSION: SnapshotVersion = Compact(3);
 
 /// The snapshot that we store on disk.
 #[derive(Decode, Encode)]
-struct Snapshot<B: BlockT> {
+struct Snapshot<H: HashOutput> {
```
Comment (ggwpez) [inline substrate/utils/frame/remote-externalities/src/lib.rs:66]: Is it not possible to remove the bound here?

File: prdoc/pr_3953.prdoc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    description: |
+      Modify `RemoteExternalities`, `Mode`, `OnlineConfig` and`Snapshot` to rely now on generic parameter, instead of `BlockT`
+      Modify in consequence their implementation to be used with `Hash` or `HashOutput`
+      Adapt Builder struct and implementation for these new bounds
+
+crates: 
+- name: frame-remote-externalities
+  bump: major
+- name: pallet-state-trie-migration
+  bump: major
```
Comment (ggwpez) [inline prdoc/pr_3953.prdoc:17]:
```suggestion
  bump: patch
```
Looks like the only change in that crate was `run_with_limits`, and it is private.

File: prdoc/pr_3953.prdoc [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+- name: try-runtime-cli
+  bump: major
```
Comment (ggwpez) [inline prdoc/pr_3953.prdoc:19]:
```suggestion
  bump: patch
```
ditto

File: prdoc/pr_3953.prdoc [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Adapt RemoteExternalities and its related types to be used with generic hash parameters
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      Modify `RemoteExternalities`, `Mode`, `OnlineConfig` and`Snapshot` to rely now on generic parameter, instead of `BlockT`
+      Modify in consequence their implementation to be used with `Hash` or `HashOutput`
+      Adapt Builder struct and implementation for these new bounds
```
Comment (ggwpez) [inline prdoc/pr_3953.prdoc:11]:
```suggestion
      Adapt Builder struct and implementation for these new bounds.
```
Punctuation please.

File: substrate/utils/frame/remote-externalities/src/lib.rs [hunk 5]
```diff
@@ -240,20 +240,20 @@ pub struct OnlineConfig<B: BlockT> {
 	pub hashed_keys: Vec<Vec<u8>>,
 }
 
-impl<B: BlockT> OnlineConfig<B> {
+impl<H: HashOutput> OnlineConfig<H> {
```
Comment (ggwpez) [inline substrate/utils/frame/remote-externalities/src/lib.rs:243]: Looks like only a `Clone` is needed here.

File: substrate/utils/frame/remote-externalities/src/lib.rs [hunk 6]
```diff
@@ -240,20 +240,20 @@ pub struct OnlineConfig<B: BlockT> {
...
 	}
 
-	fn at_expected(&self) -> B::Hash {
+	fn at_expected(&self) -> H {
 		self.at.expect("block at must be initialized; qed")
 	}
 }
 
-impl<B: BlockT> Default for OnlineConfig<B> {
+impl<H: HashOutput> Default for OnlineConfig<H> {
```
Comment (ggwpez) [inline substrate/utils/frame/remote-externalities/src/lib.rs:256]: How is the bound needed here?

File: substrate/utils/frame/remote-externalities/src/lib.rs [hunk 7]
```diff
@@ -267,7 +267,7 @@ impl<B: BlockT> Default for OnlineConfig<B> {
 	}
 }
 
-impl<B: BlockT> From<String> for OnlineConfig<B> {
+impl<H: HashOutput> From<String> for OnlineConfig<H> {
```
Comment (ggwpez) [inline substrate/utils/frame/remote-externalities/src/lib.rs:270]: And here?

**Review by ggwpez — APPROVED** [review]
Some nits but looks good 👍

File: substrate/utils/frame/remote-externalities/src/lib.rs [hunk 8]
```diff
@@ -240,20 +240,20 @@ pub struct OnlineConfig<B: BlockT> {
...
 	fn rpc_client(&self) -> &HttpClient {
 		self.transport
 			.as_client()
 			.expect("http client must have been initialized by now; qed.")
 	}
 
-	fn at_expected(&self) -> B::Hash {
-		self.at.expect("block at must be initialized; qed")
+	fn at_expected(&self) -> H {
+		self.clone().at.expect("block at must be initialized; qed")
```
Comment (lean-apple) [inline substrate/utils/frame/remote-externalities/src/lib.rs:252]: @ggwpez not sure about this change

Comment (ggwpez) [reply substrate/utils/frame/remote-externalities/src/lib.rs:252]: Yes it is more correct than previous - which pulled in `Copy`. I think its good 👍

Comment (bkchr) [inline substrate/utils/frame/remote-externalities/src/lib.rs:252]:
```suggestion
		self.at.clone().expect("block at must be initialized; qed")
```

File: prdoc/pr_3953.prdoc [hunk 9]
```diff
@@ -0,0 +1,19 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Adapt RemoteExternalities and its related types to be used with generic hash parameters
+
+doc:
+  - audience: Runtime Dev
```
Comment (bkchr) [inline prdoc/pr_3953.prdoc:7]:
```suggestion
  - audience: Node Dev
```

**Review by bkchr — APPROVED** [review]

**Pull request merged**